crates/sui-sdk/examples/json_rpc_errors.rs

```diff
@@ -0,0 +1,26 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+mod utils;
+use sui_sdk::error::{Error, JsonRpcError};
+use utils::setup_for_read;
+
+#[tokio::main]
+async fn main() -> Result<(), anyhow::Error> {
+    let (sui, active_address) = setup_for_read().await?;
+    let coin_type = Some("0x42".to_string());
+    let coins = sui
+        .coin_read_api()
+        .get_coins(active_address, coin_type.clone(), None, Some(5))
+        .await;
+    let error = coins.unwrap_err();
+    if let Error::RpcError(rpc_error) = error {
+        let converted: JsonRpcError = rpc_error.into();
+        println!(" *** RpcError ***");
+        println!("{converted}");
+        println!("{}", converted.is_client_error());
+    } else {
+        panic!("Expected Error::RpcError, got {:?}", error);
```

Comment:
Might be better to use `bail!("Expected Error::RpcError, got {:?}", error)` instead of panic.

Reply:
ok, will modify

Reply:
Actually... in an example (or test code) it's better to use `panic!` rather than `bail!` because `bail!` will not help you get a stacktrace.  I realise that I'm imparting these rules about `Result` vs `panic!` piecemeal, koan style, so here's the full set of "Ashok's mental model of when to panic":


#### Services

**Never panic** (mostly).  The one exception to this rule is during initialisation (i.e. if you can't bring up the service fully, just panic, don't try and recover and then "half serve" requests).  This means all failures must be recorded as `Result<T, E>`-s where `E` is a **structured error type**, that can be translated to an error response. The `thiserror` crate is good for defining these error types.

#### Libraries

Also heavily favour not panicking, because you don't know where your library is going to be used, and panicking is not a well encapsulated error communication strategy. `thiserror` or similar would be my goto here, to provide composable, structured errors on the library's public interface.

#### CLI

Hopefully you followed the advice of your libraries not panicking, because you also don't want to panic here.  When a CLI tool fails, we need to provide a human-readable error message with context (and not the kind of context you get with `panic!`, which is a stacktrace). `anyhow` is a good crate for boxed errors: An `anyhow::Error` can hold errors of any structured error type (anything that implements `Error`, which `thiserror` has a `derive` for), and also allows you to annotate error results with context, so that you can explain where in the operation things went wrong (without resorting to a stacktrace).

#### Tests (and Examples)

Panic early, panic often, panic with reckless abandon (as long as we're talking purely about test or example code -- i.e. it's not a library.  If so, see above).  Reason being -- the context you need when something goes wrong in a test is the stacktrace, which is what `panic!` (or operations that panic like `unwrap` or `expect`) gives you.  If you don't do this, you'll need to adopt the CLI strategy of manually reconstructing the context using an `anyhow::Error` and `context` calls, but this is usually redundant -- we don't care as much about the error being human readable.